Python 3 does not install pip

Python's package contains the following two lines:

``` bash
if spec.satisfies('@3:'):
    config_args.append('--without-ensurepip')
```

Which dates back to c65fd3a28968f549eed361d85f715caa4f8f6436. I was able to install python with pip after commenting out this line. Is this flag necessary, or could pip be a variant? Pip may be another package manager, but it's not superfluous.

> using this [the +pip option for Python 3] as a variant shouldn't cause issues. It hasn't for me at least.

Unfortunately, this is not true anymore with Python 3.7 :/ `ensurepip` now requires `ctypes` for some reason, see https://bugs.python.org/issue31652, and `ctypes` doesn't compile the way the `python` package is currently set up, which results in this error:

```
==> Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j32' 'install'

1 error found in build log:
     8911      File "/tmp/tmp7y2lkesw/pip-10.0.1-py2.py3-none-any.whl/pip/_internal/index.py", line 25, in <module>
     8912      File "/tmp/tmp7y2lkesw/pip-10.0.1-py2.py3-none-any.whl/pip/_internal/download.py", line 39, in <module>
     8913      File "/tmp/tmp7y2lkesw/pip-10.0.1-py2.py3-none-any.whl/pip/_internal/utils/glibc.py", line 3, in <module>
     8914      File "/cnk/common/tools/spack/var/spack/stage/python-3.7.0-wr7smr6tublir72kp4koxj33zrmrv6fk/Python-3.7.0/Lib/ctypes/__init__.py", line 7, in <module>
     8915        from _ctypes import Union, Structure, Array
     8916    ModuleNotFoundError: No module named '_ctypes'
  >> 8917    make: *** [install] Error 1
```

I've tried adding `libffi` as a dependency, as recommended on the Python bug tracker, but it still fails, so more digging is needed, which I don't have time for right now...

--- a/var/spack/repos/builtin/packages/py-build/package.py
+++ b/var/spack/repos/builtin/packages/py-build/package.py
@@ -16,7 +16,7 @@ class PyBuild(PythonPackage):
 
     variant('virtualenv', default=False, description='Install optional virtualenv dependency')
 
-    depends_on('python@3.6:', type=('build', 'run'))
+    depends_on('python@3.6:+ensurepip', type=('build', 'run'))
     depends_on('py-setuptools', type='build')
     depends_on('py-packaging@19:', type=('build', 'run'))
     depends_on('py-pep517@0.9.1:', type=('build', 'run'))

--- a/var/spack/repos/builtin/packages/python/package.py
+++ b/var/spack/repos/builtin/packages/python/package.py
@@ -3,6 +3,7 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+import glob
 import json
 import os
 import platform
@@ -154,12 +155,13 @@ class Python(AutotoolsPackage):
     variant('nis',      default=False, description='Build nis module')
     variant('zlib',     default=True,  description='Build zlib module')
     variant('bz2',      default=True,  description='Build bz2 module')
-    variant('lzma',     default=True,  description='Build lzma module')
+    variant('lzma',     default=True,  description='Build lzma module', when='@3.3:')
     variant('pyexpat',  default=True,  description='Build pyexpat module')
     variant('ctypes',   default=True,  description='Build ctypes module')
     variant('tkinter',  default=False, description='Build tkinter module')
     variant('uuid',     default=True,  description='Build uuid module')
     variant('tix',      default=False, description='Build Tix module')
+    variant('ensurepip', default=True, description='Build ensurepip module', when='@2.7.9:2,3.4:')
 
     depends_on('pkgconfig@0.9.0:', type='build')
     depends_on('gettext +libxml2', when='+libxml2')
@@ -272,7 +274,7 @@ class Python(AutotoolsPackage):
 
         variants = ''
         for module in ['readline', 'sqlite3', 'dbm', 'nis',
-                       'zlib', 'bz2', 'lzma', 'ctypes', 'uuid']:
+                       'zlib', 'bz2', 'ctypes', 'uuid']:
             try:
                 python('-c', 'import ' + module, error=os.devnull)
                 variants += '+' + module
@@ -294,7 +296,23 @@ class Python(AutotoolsPackage):
         except ProcessError:
             variants += '~pyexpat'
 
-        # Some modules changed names in Python 3
+        # Some modules are version-dependent
+        if Version(version_str) >= Version('3.3'):
+            try:
+                python('-c', 'import lzma', error=os.devnull)
+                variants += '+lzma'
+            except ProcessError:
+                variants += '~lzma'
+
+        if Version(version_str) in ver('2.7.9:2,3.4:'):
+            # The ensurepip module is always available, but won't work if built with
+            # --without-ensurepip. A more reliable check to see if the package was built
+            # with --with-ensurepip is to check for the presence of a pip executable.
+            if glob.glob(os.path.join(os.path.dirname(exes[0]), 'pip*')):
+                variants += '+ensurepip'
+            else:
+                variants += '~ensurepip'
+
         if Version(version_str) >= Version('3'):
             try:
                 python('-c', 'import tkinter', error=os.devnull)
@@ -463,8 +481,11 @@ class Python(AutotoolsPackage):
                 raise ValueError(
                     '+ucs4 variant not compatible with Python 3.3 and beyond')
 
-        if spec.satisfies('@3:'):
-            config_args.append('--without-ensurepip')
+        if spec.satisfies('@2.7.9:2,3.4:'):
+            if '+ensurepip' in spec:
+                config_args.append('--with-ensurepip')
+            else:
+                config_args.append('--without-ensurepip')
 
         if '+pic' in spec:
             cflags.append(self.compiler.cc_pic_flag)
@@ -640,6 +661,10 @@ class Python(AutotoolsPackage):
                     self.command('-c', 'import tkinter.tix')
                 else:
                     self.command('-c', 'import Tix')
+
+            # Ensure that ensurepip module works
+            if '+ensurepip' in spec:
+                self.command('-c', 'import ensurepip')
 
     # ========================================================================
     # Set up environment to make install easy for python extensions.

